app/models/topic_tracking_state.rb

```diff
@@ -281,73 +331,120 @@ def self.report(user, topic_id = nil)
       muted_tag_ids: tag_ids
     )
 
+    refine_sql_proc = Proc.new do |opts|
+      @refine_proc_called = true
+      sql << "\nUNION ALL\n\n" + report_raw_sql(
+        select: opts[:select],
+        topic_id: topic_id,
+        skip_new: true,
+        skip_unread: true,
+        skip_order: true,
+        staff: user.staff?,
+        admin: user.admin?,
+        user: user,
+        muted_tag_ids: tag_ids,
+        custom_state_filter: opts[:custom_state_filter]
+      )
+    end
+
+    refinement_called = false
+    (@refine_methods || []).each do |refinement|
+      refinement.call(user, tag_ids, topic_id, refine_sql_proc)
+
+      # all SQL added to this query should go through refine_sql_proc, which
+      # will append to the SQL itself (there is no other way to append SQL then)
+      raise RefinementProcNotCalledError if !@refine_proc_called
+
+      refinement_called = true
+      @refine_proc_called = false
+    end
+
     if SiteSetting.tagging_enabled && TopicTrackingState.include_tags_in_report?
       sql = <<~SQL
-        WITH X AS (#{sql})
+        WITH tags_included_cte AS (
+          #{sql}
+        )
         SELECT *, (
           SELECT ARRAY_AGG(name) from topic_tags
              JOIN tags on tags.id = topic_tags.tag_id
-             WHERE topic_id = X.topic_id
+             WHERE topic_id = tags_included_cte.topic_id
           ) tags
-        FROM X
+        FROM tags_included_cte
       SQL
     end
 
-    DB.query(
+    report = DB.query(
       sql,
         user_id: user.id,
         topic_id: topic_id,
         min_new_topic_date: Time.at(SiteSetting.min_new_topics_time).to_datetime
     )
+
+    # we do not want duplicate topic ids in the
+    # topic tracking state
+    if refinement_called
+      report = report.uniq do |tracking_state|
+        tracking_state.topic_id
+      end
+    end
+
+    report
```

Comment:
I wonder if we should just do a `report.limit(4000)` here to avoid too much state craziness? Maybe even less?

Reply:
Yeah I think some safety is in order, I think a hard limit of 5000 on refinements is a good start. as a constant.